_posts/2020-10-21-explicit-term-inference-in-scala-3.md

```diff
@@ -0,0 +1,433 @@
+You do not need to port your codebase right away, it can be an incremental and gradual process.
+
+## Extension methods
+
+When you do not have control over a type but you need to extend its capabilities with a new method,
+Scala 3 allows you to define an *extension method*
+
+Assume that you are working with `List[Try[String]]` and that you often need to retrieve
+the elements for which the computation succeeded.
+
+Then you can extend this type to have a `collectSucceded` method:
+
+```scala
+// ListTryOps.scala
+
+import scala.util.{Try, Success}
+
+extension (ls: List[Try[String]]) def collectSucceeded: List[String] =
+  ls.collect { case Success(x) => x }
+```
+To remember the syntax, remark how `collectSucceeded` follows the object on which it will be available: `ls.collectSucceeded`.
+Note that extension methods can have type parameters as well:
+
+```scala
+extension[A] (ls: List[Try[A]]) def collectSucceeded: List[A] =
+  ls.collect { case Success(x) => x }
+```
+
+Finally, you can add several methods without repeating the `extension` declaration:
+
+```scala
+extension[A] (ls: List[Try[A]]):
+  def collectSucceeded: List[A] =
+    ls.collect { case Success(x) => x }
+  def getIndexOfFirstFailure: Option[Int] =
+    ls.zipWithIndex.find((t, _) => t.isFailure)
+      .map(_._2)
+```
+
+Extensions can be imported by name or using the `_` wildcard:
+
+```scala
+// Main.scala
+import scala.util.Try
+// import ListTryOps._ // Using wildcard
+import ListTryOps.collectSucceeded
+
+def getLastTweet(username: String): Try[String] = ???
+
+@main def main =
+  val niceTweets = List(getLastTweet("scala_lang"), getLastTweet("odersky")).collectSucceeded
+  println(niceTweets)
+```
+
+### How was this done in Scala 2?
+
+This pattern was particularly cumbersome to implement prior to Scala 3.
+A typical approach would be the following:
+
+```scala
+implicit class ListTryExtension[A](val in: List[Try[A]]) extends AnyVal {
+  def collectSucceeded: List[A] = ls.collect { case Success(a) => a }
+  def getIndexOfFirstFailure: Option[Int] =
+      in
+        .zipWithIndex.find { case (t, _) => t.isFailure }
+        .map { case (_, index) => index }
+}
+```
+Remark that you need to define a name for the class, even if this is probably never going to be used besides the import statement.
+You also need to understand what `AnyVal` is, why it is good practice to extend it and
+what its limitations are.
+
+Some experience is required when reading this code to understand that its only goal is to
+add a couple of methods to `List[Try[A]]`.
+
+### Find out more
+
+You can find more information about extension methods on [the dedicated documentation page](http://dotty.epfl.ch/docs/reference/contextual/extension-methods.html).
+I also suggest that you read how they complement another new Scala 3 feature: [opaque types](https://dotty.epfl.ch/docs/reference/other-new-features/opaques.html).
```

Comment:
`We` is used everywhere except here and a couple of other places (I'll mark them), so maybe replace `I` with `we`.